checkmarx-ast-eclipse-plugin/src/com/checkmarx/eclipse/views/actions/ActionFilterStatePreference.java

```diff
@@ -0,0 +1,143 @@
+package com.checkmarx.eclipse.views.actions;
+
+import org.eclipse.jface.action.Action;
+import org.eclipse.jface.action.IAction;
+import org.eclipse.jface.action.IMenuCreator;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.graphics.Rectangle;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Listener;
+import org.eclipse.swt.widgets.Menu;
+import org.eclipse.swt.widgets.MenuItem;
+import org.eclipse.swt.widgets.ToolItem;
+
+import com.checkmarx.eclipse.Activator;
+import com.checkmarx.eclipse.enums.ActionName;
+import com.checkmarx.eclipse.enums.PluginListenerType;
+import com.checkmarx.eclipse.enums.State;
+import com.checkmarx.eclipse.views.DataProvider;
+import com.checkmarx.eclipse.views.PluginListenerDefinition;
+import com.checkmarx.eclipse.views.filters.FilterState;
+import com.google.common.eventbus.EventBus;
+
+class ActionFilterStatePreference extends Action implements IMenuCreator
+{
+  private Menu menu;
+  private static final String ACTION_FILTER_STATE_TOOLTIP = "State";
+  private static final String ACTION_FILTER_STATE_ICON_PATH = "/icons/filter_ps.png";
+  private EventBus pluginEventBus;
+  
+  
+  public static final String FILTER_NOT_EXPLOITABLE = "Not Exploitable";
+  public static final String FILTER_CONFIRMED = "Confirmed";
+  public static final String FILTER_PROPOSED_NON_EXPLOITABLE = "Proposed Not Exploitable";
+  public static final String FILTER_URGENT = "Urgent";
+  public static final String FILTER_IGNORED = "Ignored";
+  public static final String FILTER_NOT_IGNORED = "Not Ignored";
+  public static final String FILTER_TO_VERIFY = "To Verify";
+
+  ActionFilterStatePreference(EventBus pluginEventBus)
+  {
+    super(ACTION_FILTER_STATE_TOOLTIP);
+    setId(ActionName.FILTER_CHANGED.name());
+    setImageDescriptor(Activator.getImageDescriptor(ACTION_FILTER_STATE_ICON_PATH));
+    this.pluginEventBus = pluginEventBus;
+    setMenuCreator(this);
+  }
+
+
+  @Override
+  public void dispose()
+  {
+    if (menu != null) {
+      menu.dispose();
+      menu = null;
+    }
+  }
+
+
+  @Override
+  public void runWithEvent(final Event event)
+  {
+    if (event.widget instanceof ToolItem)
+      {
+        final ToolItem toolItem = (ToolItem)event.widget;
+        final Control control = toolItem.getParent();
+        final Menu menu = getMenu(control);
+
+        final Rectangle bounds = toolItem.getBounds();
+        final Point topLeft = new Point(bounds.x, bounds.y + bounds.height);
+        menu.setLocation(control.toDisplay(topLeft));
+        menu.setVisible(true);
+      }
+  }
+
+  @Override
+  public Menu getMenu(final Control parent)
+  {
+    if (menu != null) {
+      menu.dispose();
+    }
+
+    menu = new Menu(parent);
```

Comment:
can we extract this to a method ? just because we repeating the Menuitem create.